[SPARK-23977][SQL] Support High Performance S3A committers [test-hadoop3.2]

steveloughran opened the pull request:
## What changes were proposed in this pull request?

This patch adds the binding classes to enable spark to switch dataframe output to using the S3A zero-rename committers shipping in Hadoop 3.1+. It adds a source tree into the hadoop-cloud-storage module which only compiles with the hadoop-3.2 profile, and contains a binding for normal output and a specific bridge class for Parquet (as the parquet output format requires a subclass of `ParquetOutputCommitter`.

## How was this patch tested?

Commit algorithms are a critical topic. There's no formal proof of correctness, but the algorithms are documented an analysed in [A Zero Rename Committer](https://github.com/steveloughran/zero-rename-committer/releases). This also reviews the classic v1 and v2 algorithms, IBM's swift committer and the one from EMRFS which they admit was based on the concepts implemented here.

Test-wise

* There's a public set of scala test suites [on github](https://github.com/hortonworks-spark/cloud-integration)
* We have run integration tests against Spark on Yarn clusters.
* This code has been shipping for ~12 months in HDP-3.x.

steveloughran committed:
[SPARK-23977][SQL]  Add a binding to the Hadoop 3.1 Zero-rename committer for S3A

These committers for dataframes writing through the Hadoop Map Reduce committer API delivers high performance job and task commits, while guaranteeing that the output of individual tasks are not visible in the output until job commit, and that a failure during task commit does not leave the destination in an unknown state. (side note: The EMR FS optimized committer inspired this work doesn't offer the latter guarantee).

Parquet support comes through an extra binding class which satisfies the Parquet library's requirement that all committers must be a subclass of ParquetOutputCommitter by bridging to the new Hadoop MR filesystem-specific factory mechanism.

This patch doesn't include functional tests of the committers, as that requires full AWS credentials and test setup to use them.

Change-Id: Icae6dc9ef3075d5c86e256edddd6df24c2c2f9f9

steveloughran committed:
SPARK-23677: add docs, review javadocs

This patch updates the cloud-integration document with details on the new committer, along with more criticism of the existing committers, and a summary of the performance and safety aspects of the store.

Change-Id: If1bea909d2e4a419e58dc084f127328c7dc0f46f

steveloughran commented:
@felixcheung -updated PR here. If could d/l and test that'd be really great

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/12114/
Test PASSed.

SparkQA commented:
**[Test build #106908 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/106908/testReport)** for PR 24970 at commit [`b8307f4`](https://github.com/apache/spark/commit/b8307f494543d2f12b436a86bf392faef3792214).

SparkQA commented:
**[Test build #106908 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/106908/testReport)** for PR 24970 at commit [`b8307f4`](https://github.com/apache/spark/commit/b8307f494543d2f12b436a86bf392faef3792214).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/106908/
Test PASSed.

steveloughran commented:
build/test still isn't using hadoop-3.2 profile (else CommitterBindingSuite would be in the test results). Is there a way to force this?

vanzin commented on docs/cloud-integration.md:
```diff
@@ -125,12 +125,13 @@ consult the relevant documentation.
 
 For object stores whose consistency model means that rename-based commits are safe
-use the `FileOutputCommitter` v2 algorithm for performance:
+use the `FileOutputCommitter` v2 algorithm for performance, v1 for safety
```
You're missing some punctuation at the end of almost all paragraphs you're touching.

steveloughran replied:
will review and correct

vanzin commented on docs/cloud-integration.md:
```diff
@@ -143,8 +144,34 @@ job failure:
+| Amazon S3     | S3A       | Unsafe                  | O(data) |
+| Azure Storage | wasb      | Safe                    | O(files) |
+| Azure Datalake Gen 2 | abfs | Safe                  | O(1) |
+| Google GCS    | gs        | Saf  e                  | O(1) |
```
Safe

vanzin commented on docs/cloud-integration.md:
```diff
@@ -190,6 +217,46 @@ while they are still being written. Applications can write straight to the monit
+
+More details on these committers can be found in the latest Hadoop documentation.
+
+Amazon EMR's S3 connector also has a zero-rename committer. Consult
```
I'd just not mention this. (They're free to contribute their committer to the community to earn the mention.)

steveloughran replied:
cut

vanzin commented on docs/cloud-integration.md:
````diff
@@ -190,6 +217,46 @@ while they are still being written. Applications can write straight to the monit
+spark.sql.parquet.output.committer.class org.apache.spark.internal.io.cloud.BindingParquetOutputCommitter
+```
+
+Normal dataframe write commands will then use this committer for any format
````
This paragraph sounds a little self-contradictory.

- Committer will be used for any format that doesn't have a custom one
- These are the supported formats.

Does that mean that formats that aren't explicitly called out are not necessarily supported or tested or what?

(Removing the list of "supported formats" would help in avoiding the ambiguity.)

steveloughran replied:
changed to say : "output formats which are known to work (i.e. the ones tested)"

vanzin replied:
I think that phrasing just raises question of "which are these formats". If you really insist on saying something, just say "tested with the most common formats supported by Spark".

steveloughran replied:
Done

vanzin commented on hadoop-cloud/pom.xml:
```diff
@@ -200,6 +200,46 @@
+
+      <build>
+        <plugins>
+          <!-- Include a source dir for Hadoop 3.2+ only -->
```
Unnecessary comment (this is under the profile after all).

steveloughran replied:
ok

vanzin commented on hadoop-cloud/src/hadoop-3/main/scala/org/apache/spark/internal/io/cloud/BindingParquetOutputCommitter.scala:
```diff
@@ -0,0 +1,124 @@
+
+  logDebug(s"${this.getClass.getName} binding to configured PathOutputCommitter and dest $path")
+
+  val committer = new BindingPathOutputCommitter(path, context)
```
private?

steveloughran replied:
done

vanzin commented on hadoop-cloud/src/hadoop-3/main/scala/org/apache/spark/internal/io/cloud/BindingParquetOutputCommitter.scala:
```diff
@@ -0,0 +1,124 @@
+   * This is the committer ultimately bound to.
+   * @return the committer instantiated by the factory.
+   */
+  def boundCommitter(): PathOutputCommitter = {
```
Is this useful outside of testing? If not, then `private[cloud]`.

steveloughran replied:
its only for testing the transitive binding -marked as private[cloud]

vanzin commented on hadoop-cloud/src/hadoop-3/main/scala/org/apache/spark/internal/io/cloud/BindingParquetOutputCommitter.scala:
```diff
@@ -0,0 +1,124 @@
+    committer.getCommitter()
+  }
+
+  override def getWorkPath: Path = {
```
add `()`

steveloughran replied:
done, and turned the IDE off from hinting this

vanzin commented on hadoop-cloud/src/hadoop-3/main/scala/org/apache/spark/internal/io/cloud/BindingParquetOutputCommitter.scala:
```diff
@@ -0,0 +1,124 @@
+   * @param jobContext job context
+   * @param state final state of the job
+   */
+  override def abortJob(
```
Declaration fits in one line.

vanzin commented on hadoop-cloud/src/hadoop-3/main/scala/org/apache/spark/internal/io/cloud/PathOutputCommitProtocol.scala:
```diff
@@ -0,0 +1,281 @@
+ * API to create the committer.
+ * This is what [[org.apache.hadoop.mapreduce.lib.output.FileOutputFormat]] does,
+ * but as [[HadoopMapReduceCommitProtocol]] still uses the original
+ * `org.apache.hadoop.mapred.FileOutputFormat` binding
```
I'm having a hard time parsing this paragraph past this point.

steveloughran replied:
I just cut the whole para; it doesnt add much

vanzin commented on hadoop-cloud/src/hadoop-3/main/scala/org/apache/spark/internal/io/cloud/PathOutputCommitProtocol.scala:
```diff
@@ -0,0 +1,281 @@
+ * subclasses do not do this, overrides those subclasses to using the
+ * factory mechanism now supported in the base class.
+ *
+ * In `setupCommitter` the factory is bonded to and the committer for
```
"bonded to ..."? Seems to be missing something.

vanzin commented on hadoop-cloud/src/hadoop-3/main/scala/org/apache/spark/internal/io/cloud/PathOutputCommitProtocol.scala:
```diff
@@ -0,0 +1,281 @@
+  require(destination != null, "Null destination specified")
+
+  /** The destination path. This is serializable in Hadoop 3. */
+  val destPath: Path  = new Path(destination)
```
Too many spaces before `=`. Also, private?

steveloughran replied:
done

vanzin commented on hadoop-cloud/src/hadoop-3/main/scala/org/apache/spark/internal/io/cloud/PathOutputCommitProtocol.scala:
```diff
@@ -0,0 +1,281 @@
+    s" destination=$destPath;" +
+    s" dynamicPartitionOverwrite=$dynamicPartitionOverwrite")
+
+  if (dynamicPartitionOverwrite) {
```
Maybe move this earlier (e.g. before initializing fields and etc)?

steveloughran replied:
gladly

vanzin commented on hadoop-cloud/src/hadoop-3/main/scala/org/apache/spark/internal/io/cloud/PathOutputCommitProtocol.scala:
```diff
@@ -0,0 +1,281 @@
+    throw new IOException(PathOutputCommitProtocol.UNSUPPORTED)
+  }
+
+  def getDestination(): String = destination
```
If only used for testing, use `private[cloud]`.

Or maybe even make `destination` a `private[cloud] val`.

steveloughran replied:
removed

vanzin commented on hadoop-cloud/src/hadoop-3/main/scala/org/apache/spark/internal/io/cloud/PathOutputCommitProtocol.scala:
```diff
@@ -0,0 +1,281 @@
+   *         [[PathOutputCommitter]].
+   */
+  override protected def setupCommitter(
+      context: TaskAttemptContext): PathOutputCommitter = {
```
Fits in one line.

vanzin commented on hadoop-cloud/src/hadoop-3/main/scala/org/apache/spark/internal/io/cloud/PathOutputCommitProtocol.scala:
```diff
@@ -0,0 +1,281 @@
+      ext: String): String = {
+
+    val workDir = committer.getWorkPath
+    val parent = dir.map(d => new Path(workDir, d)).getOrElse(workDir)
```
`.map { d => ... }`

steveloughran replied:
done

vanzin commented on hadoop-cloud/src/hadoop-3/main/scala/org/apache/spark/internal/io/cloud/PathOutputCommitProtocol.scala:
```diff
@@ -0,0 +1,281 @@
+   * @return an absolute path
+   */
+  override def newTaskTempFileAbsPath(
+    taskContext: TaskAttemptContext,
```
nit: double indent for params

steveloughran replied:
done here and in other uses

vanzin commented on hadoop-cloud/src/hadoop-3/main/scala/org/apache/spark/internal/io/cloud/PathOutputCommitProtocol.scala:
```diff
@@ -0,0 +1,281 @@
+  }
+
+  /**
+   * Absolute files are still renamed into place with a warning.
```
nit: this should explain what this method does.

vanzin commented on hadoop-cloud/src/hadoop-3/main/scala/org/apache/spark/internal/io/cloud/PathOutputCommitProtocol.scala:
```diff
@@ -0,0 +1,281 @@
+   * @return a name for a file which must be unique across all task attempts
+   */
+  protected def buildFilename(
+    taskContext: TaskAttemptContext,
```
Fits in one line.

vanzin commented on hadoop-cloud/src/hadoop-3/main/scala/org/apache/spark/internal/io/cloud/PathOutputCommitProtocol.scala:
```diff
@@ -0,0 +1,281 @@
+
+  override def commitTask(
+      taskContext: TaskAttemptContext): FileCommitProtocol.TaskCommitMessage = {
+      logDebug("Commit task")
```
wrong indent. But this whole method seems unnecessary.

vanzin commented on hadoop-cloud/src/hadoop-3/main/scala/org/apache/spark/internal/io/cloud/PathOutputCommitProtocol.scala:
```diff
@@ -0,0 +1,281 @@
+  override def commitJob(
+      jobContext: JobContext,
+      taskCommits: Seq[FileCommitProtocol.TaskCommitMessage]): Unit = {
+    logDebug(s"commit job with ${taskCommits.length} task commit message(s)")
```
Are these overrides just to add a debug message really helpful?

If so, why not have the debug message in the parent class?

steveloughran replied:
It is handy to have the option of debugging what is going on, and I was trying to be as minimally invasive to the existing code as possible. But given their value, pulling them up is a good idea. Will do that and then delete these methods

vanzin replied:
Did you mean you were going to change the parent class as part of this change or separately?

steveloughran replied:
pulling up. FWIW, If I were going to make bigger changes to the parent class it'd be

* to actually log the duration of renames
* consider actually doing them in parallel. Not just for performance against slow stores (s3, openstack), but to avoid renaming large files triggering timeouts. I've been playing a lot with rename perf recently (HADOOP-15183) and even with parallelized renaming of individual blocks there's scope for speedup
* Track which parent partition paths have already been created, so skip all the exists() checks. Saves 3 S3 requests per file.

That will be something for a new JIRA

vanzin commented on hadoop-cloud/src/hadoop-3/main/scala/org/apache/spark/internal/io/cloud/PathOutputCommitProtocol.scala:
```diff
@@ -0,0 +1,281 @@
+  }
+
+  /**
+   * Abort the task; log and ignore any failure thrown.
```
I understand the spirit, but if that's so important, why is it not done in the parent class? This feels like the wrong place for it.

steveloughran replied:
I think it should be in the parent. It surfaces with these committers more because 

1. the primary failure of an abort task in a classic cluster FS is that the cluster FS is unreachable, which is traumatic enough more things fail.
1. with alternative cloud stores and the like, there are different ways for things to go wrong. For example, the user issuing requests to AWS S3 doesn't have the "abort multipart upload" permission.

If you think it should be pulled up —which delivers consistency— then I'm happy to do this

vanzin replied:
> If you think it should be pulled up

Yes, I think that's better. That way the reasoning is consistent - preserve the original exception, always - instead of different committers behaving in different ways.

vanzin replied:
Were you planning on addressing this, or at least filing a follow-up bug?

steveloughran replied:
pulled up with separate handlers for committer abort and staging dir cleanup. Rationale: it may be that a committer fails for some reason (e.g. IAM permissions for abort) but delete() still works.

vanzin commented on hadoop-cloud/src/hadoop-3/main/scala/org/apache/spark/internal/io/cloud/package.scala:
```diff
@@ -0,0 +1,115 @@
+ *   spark.sql.sources.commitProtocolClass org.apache.spark.internal.io.cloud.PathOutputCommitProtocol
+ * }}}
+ *
+ * It can be done programmatically by calling [[cloud.bind()]] on the
```
While having a programmatic helper is nice, this is in an `internal` package, so no one will find it since those don't show up in documentation...

What I'd suggest, perhaps as a separate task, would be to automatically switch to these committers if they're available and if there's no drawback in doing so, so that by default having a build with Spark + Hadoop 3.2 + the cloud bindings automatically gives you this stuff, without needing to configure things at all. (Perhaps with a flag to disable the automatic use of the committers if someone does run into issues.)

steveloughran replied:
Let me follow that up in a separate comment in the main conversation. We can do more here

vanzin commented on hadoop-cloud/src/hadoop-3/test/scala/org/apache/spark/internal/io/cloud/CommitterBindingSuite.scala:
```diff
@@ -0,0 +1,153 @@
+
+  test("cloud binding to SparkConf") {
+    val sc = new SparkConf()
+    cloud.bind(sc)
```
This isn't actually testing anything...

steveloughran replied:
cut

vanzin commented on hadoop-cloud/src/hadoop-3/test/scala/org/apache/spark/internal/io/cloud/CommitterBindingSuite.scala:
```diff
@@ -0,0 +1,153 @@
+   * @return the new job
+   * @throws IOException failure
+   */
+  @throws[IOException]
```
These annotations aren't actually that helpful. Even less so in test code.

steveloughran replied:
cut

vanzin commented on hadoop-cloud/src/hadoop-3/test/scala/org/apache/spark/internal/io/cloud/CommitterBindingSuite.scala:
```diff
@@ -0,0 +1,153 @@
+   * Create a a new job. Sets the task attempt ID.
+   *
+   * @return the new job
+   * @throws IOException failure
```
These comments aren't super helpful.

steveloughran replied:
force of habit: Cut

vanzin commented on hadoop-cloud/src/hadoop-3/test/scala/org/apache/spark/internal/io/cloud/CommitterBindingSuite.scala:
```diff
@@ -0,0 +1,153 @@
+  test("CommitterSerialization") {
+    val tempDir = File.createTempFile("ser", ".bin")
+
+    tempDir.delete();
```
no semi-colons.

steveloughran replied:
again, habit, again, cut

vanzin commented on hadoop-cloud/src/hadoop-3/test/scala/org/apache/spark/internal/io/cloud/CommitterBindingSuite.scala:
```diff
@@ -0,0 +1,153 @@
+    val tempDir = File.createTempFile("ser", ".bin")
+
+    tempDir.delete();
+    val committer = new PathOutputCommitProtocol(jobId, tempDir.toURI.toString,false)
```
Space after `,`.

steveloughran replied:
fixed

vanzin commented on hadoop-cloud/src/hadoop-3/test/scala/org/apache/spark/internal/io/cloud/CommitterBindingSuite.scala:
```diff
@@ -0,0 +1,153 @@
+
+      val committer2 = result.asInstanceOf[PathOutputCommitProtocol]
+
+      assert(committer.getDestination() === committer2.getDestination,
```
either use `()` everywhere or nowhere... (but one of my previous comments would change this anyway.)

steveloughran replied:
removed

vanzin commented on hadoop-cloud/src/hadoop-3/test/scala/org/apache/spark/internal/io/cloud/CommitterBindingSuite.scala:
```diff
@@ -0,0 +1,153 @@
+      jobId, "file:///tmp", false)
+
+    val protocol = instance.asInstanceOf[PathOutputCommitProtocol]
+    assert("file:///tmp"=== protocol.getDestination())
```
Space before `===`.

vanzin commented on hadoop-cloud/src/hadoop-3/test/scala/org/apache/spark/internal/io/cloud/StubPathOutputCommitter.scala:
```diff
@@ -0,0 +1,115 @@
+}
+
+object StubPathOutputCommitterFactory {
+  val Name: String = "org.apache.spark.internal.io.cloud.StubPathOutputCommitterFactory"
```
Is this helpful? What's wrong with `classOf[StubPathOutputCommitterFactory].getName()`?

vanzin commented on hadoop-cloud/src/main/scala/org/apache/spark/internal/io/cloud/PathCommitterConstants.scala:
```diff
@@ -0,0 +1,87 @@
+   * output committer factory to use unless there is a specific
+   * one for a schema.
+   */
+  val OUTPUTCOMMITTER_FACTORY_CLASS = "mapreduce.pathoutputcommitter.factory.class"
```
This and a bunch of other constants in this class don't seem to be used anywhere.

steveloughran replied:
I'm using a copy of it in my downstream integration tests; they are really useful for people trying to play with this stuff.

the alternative is "document well.

what would you prefer

vanzin replied:
I'd prefer to not have constants that are not used. Especially since they're referencing things that are defined in other projects...

squito replied:
I agree with marcelo

steveloughran replied:
Cut all of the external ones not in use

vanzin commented on hadoop-cloud/src/main/scala/org/apache/spark/internal/io/cloud/PathCommitterConstants.scala:
```diff
@@ -0,0 +1,87 @@
+   * This is the "Pending" directory of the FileOutputCommitter;
+   * data written here is, in that algorithm, renamed into place.
+   */
+  val TEMP_DIR_NAME = "_temporary"
```
This one is used, but only in test code, so could be moved there instead.

vanzin commented on hadoop-cloud/src/main/scala/org/apache/spark/internal/io/cloud/PathCommitterConstants.scala:
```diff
@@ -0,0 +1,87 @@
+  val SUCCESS_FILE_NAME = "_SUCCESS"
+
+  /** hadoop-mapreduce option to enable the _SUCCESS marker. */
+  val CREATE_SUCCESSFUL_JOB_OUTPUT_DIR_MARKER = "mapreduce.fileoutputcommitter.marksuccessfuljobs"
```
Same thing, only used in test code.

vanzin commented on hadoop-cloud/src/test/scala/org/apache/spark/internal/io/cloud/StubPackageEntry.scala:
```diff
@@ -0,0 +1,26 @@
+package org.apache.spark.internal.io.cloud
+
+/**
+ * This class is here to make sure that the source tree is found and built.
```
Hmm, I'm not following why this class is here.

steveloughran replied:
it's a funny with maven. If you don't have a source file in the non-3.2-profile source tree, that 3.2 source tree doesn't get built. Took at while to work out and doesn't actually deliver anything of value in production code

vanzin replied:
I tried what you say locally (I deleted the `hadoop-cloud/src/test` directory), and all the hadoop-3.2 stuff was built, and tests ran.

steveloughran replied:
ok, seems to work for me too, once I can get maven past some hangings elsewhere in the build. Maybe a recent change in the compilers.

steveloughran replied:
...cutting the `test/src/scala` tree, leaving only `test/resources/log4j.properties`

vanzin reviewed (commented):
I'll trigger the tests on hadoop-3.2 (you need to change the PR title). But this could use a lot of cleanup.

vanzin commented:
retest this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/12148/
Test PASSed.

SparkQA commented:
**[Test build #106947 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/106947/testReport)** for PR 24970 at commit [`b8307f4`](https://github.com/apache/spark/commit/b8307f494543d2f12b436a86bf392faef3792214).

SparkQA commented:
**[Test build #106947 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/106947/testReport)** for PR 24970 at commit [`b8307f4`](https://github.com/apache/spark/commit/b8307f494543d2f12b436a86bf392faef3792214).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/106947/
Test PASSed.

felixcheung commented:
this is awesome and great to see! sorry, I'll have to come back a bit later to get to try this out...

vanzin commented:
Hmm, I wonder if the hadoop-cloud profile is enabled in the test scripts? Didn't see hadoop-cloud-related stuff in the build output. I'll take a look later.

vanzin commented:
Filed SPARK-28187, will post a PR shortly. Then we can test this.

vanzin commented:
#24987 let's see how that goes.

vanzin commented:
retest this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/12182/
Test PASSed.

SparkQA commented:
**[Test build #106981 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/106981/testReport)** for PR 24970 at commit [`b8307f4`](https://github.com/apache/spark/commit/b8307f494543d2f12b436a86bf392faef3792214).

SparkQA commented:
**[Test build #106981 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/106981/testReport)** for PR 24970 at commit [`b8307f4`](https://github.com/apache/spark/commit/b8307f494543d2f12b436a86bf392faef3792214).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/106981/
Test FAILed.

vanzin commented:
Alright, the build changes had the desired effect.

gaborgsomogyi commented:
@steveloughran Thanks for the effort, super useful feature!

> We have run integration tests against Spark on Yarn clusters.

Since I haven't seen streaming related tests in the shared repo I presume it has been tested with batch until now, hasn't it?

steveloughran commented:
style check: import ordering, a space (which may have gone away), and two lines in the scaladocs showing the spark-default settings. Replaced the latter with a reference to "read cloud-integration"

steveloughran commented:
I'll cut the feature (and test) and we could follow up with discussions about what to do.

Regarding automatic binding, the MRv2 APIs in Hadoop will automatically bind, so, ignoring the special case of Parquet, any configured factory for a filesystem is instantiated and used -it's up to individual factories to decide what to do. (that is: any subclass of `FileOutputFormat` which doesn't override the `getOutputCommitter()` call will go through the factories.

The key extra work in this code is to make sure that `getWorkPath()` is passed to the implementation: currently `HadoopMapReduceCommitProtocol.newTaskTempFile` casts its created committer to `FileOutputCommitter` for invoking. A big change into the new `PathOutputCommitter` interface is just to pull that method up into an interface, so that its possible to implement committers which export this feature. 

I am happy to add that interface to Hadoop 2.x and shipping 3.x versions so that even the base class could cast to PathOutputCommitter to use this. Otherwise, I could change it to switch to reflection. Key point: ignoring Parquet quirks, the `PathOutputCommitProtocol` binding could be removed entirely.


(oh, and I'm ignoring uses of the MRv1 APIs. happy to tag those as deprecated in hadoop-mapreduce.jar and help spark code migrate off them. Their time is over)

steveloughran committed:
SPARK-23977: rework after Marcelo's feedback

Change-Id: I430c50fbc0e5f5ea6aacb593808098b67d6c1bd6

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/12211/
Test PASSed.

SparkQA commented:
**[Test build #107018 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107018/testReport)** for PR 24970 at commit [`6bfea21`](https://github.com/apache/spark/commit/6bfea210bfec55d551c8c1a18e286db66b922ee4).

SparkQA commented:
**[Test build #107018 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107018/testReport)** for PR 24970 at commit [`6bfea21`](https://github.com/apache/spark/commit/6bfea210bfec55d551c8c1a18e286db66b922ee4).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/107018/
Test FAILed.

squito commented on hadoop-cloud/src/hadoop-3/main/scala/org/apache/spark/internal/io/cloud/PathOutputCommitProtocol.scala:
```diff
@@ -0,0 +1,263 @@
+        destPath,
+        context.getConfiguration)
+      logDebug(s"Using committer factory $factory")
+      committer = factory.createOutputCommitter(destPath, context)
```
what's the situation in which this does something different than `PathOutputCommitterFactory.createCommitter`?  Looks like it should be the same? https://github.com/apache/hadoop/blob/trunk/hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-core/src/main/java/org/apache/hadoop/mapreduce/lib/output/PathOutputCommitterFactory.java#L197

steveloughran replied:
It's more sometimes you want the option to say "really, really reject the classic"; a guard against that default actually happening.

The MR committer factories get to choose their own committers; currently the S3A one defaults to (still) using the rename one as switching over to a newer one. I think with HADOOP-16357 we've got the staging committer consistent with what MR expects w.r.t whether or not the destination directory is empty on startup& so for MR and spark work you should be able to switch over to it without worries. A little option here lets you fail fast on job launch to make sure that your job really has done that.

without that, you only get to find out that you are still using it in the good case (it takes a lot longer and you find that `_SUCCESS` is a 0-byte file), or the bad case (it takes a lot longer, you find that `_SUCCESS` is a 0-byte file and because of failures during task commit, your output has been silently corrupted) (*)

(*) I am of the believe that failure condition holds for FileOutputCommitter v2 and any store, it's just that for S3 and other slow-to-rename stores the window of vulnerability is a lot longer, and, without heartbeats coming back, there's a risk of the rename actually triggering a timeout. Never heard of that with Spark, but it has been known to happen with Distcp of a many GB file

steveloughran commented on docs/cloud-integration.md:
````diff
@@ -190,6 +213,42 @@ while they are still being written. Applications can write straight to the monit
+```
+
+Normal dataframe write commands will then use this committer for any format
+which does not its own custom committer. Output formats which are known
````
added "have". Also capitalized ORC and CSV & the first letters of Parquet and Avro

steveloughran committed:
SPARK-23977: updated docs, FileCommit Protocol and removed the StubPackageEntry

verified that with the hadoop-3.2 profile the new tests work

Change-Id: I0cfc39787b767182f078f883561835966dff65a2

SparkQA commented:
**[Test build #108471 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/108471/testReport)** for PR 24970 at commit [`668b15f`](https://github.com/apache/spark/commit/668b15f162d572d3f8a69cb5623a80612c1ea3d6).

SparkQA commented:
**[Test build #108471 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/108471/testReport)** for PR 24970 at commit [`668b15f`](https://github.com/apache/spark/commit/668b15f162d572d3f8a69cb5623a80612c1ea3d6).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `abstract class FileCommitProtocol extends Logging `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/108471/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/13565/
Test PASSed.

squito commented on hadoop-cloud/src/hadoop-3/main/scala/org/apache/spark/internal/io/cloud/package.scala:
```diff
@@ -0,0 +1,98 @@
+import org.apache.spark.sql.internal.SQLConf
+
+/**
+ * Package object to assist in switching to the Hadoop Hadoop 3
```
typo: repeated "Hadoop"

steveloughran replied:
fixed

squito commented on hadoop-cloud/src/hadoop-3/main/scala/org/apache/spark/internal/io/cloud/PathOutputCommitProtocol.scala:
```diff
@@ -0,0 +1,263 @@
+   * commit into the absolute path specified.
+   *
+   * The created file will be committed through the `rename()` operation
+   * of the filesystem/filestore connector, which may be a slow `O(data)`
```
if I understand this right, then even after this change, all writes into a partitioned table will still require a slow rename?  Seems that way from my quick reading of this part of spark's code:

https://github.com/apache/spark/blob/b8e13b0aeae4a25e54a4c4b8a97bc2c19edd37bb/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormatDataWriter.scala#L228-L235

is handling that not possible with spark's internal protocol?  Its fine we don't handle that for now, but should call it out in commit msg & file a follow up jira

steveloughran replied:
this is only for those cases where the protocol wants an absolute path. Traditionally spark cast the committer to a FileOutputCommitter and used methods there. Rather than make changes to what is a scarily complex piece of critical code to support new plugin committers, I just pulled up the relevant method into a new interface which can be used instead. 
Normally work goes into the subdirectory of the output, and even when partitioned there all is well. The PartitionedCommitter Ryan Blue wrote even does conflict handling within the partitions, so you can run a job targeting an existing output directory, and have it address conflict not at the base "fail if dest exists", but act on the final leaves of the partition with "rm * in that final dir", "fail if that final dir isn't empty" or "just add things relying on UUID insertion to avoid conflict and S3 update inconsistencies"

Happy to explore this more in future. FWIW, I think spark should be able to query a committer to see if could recover from a failure/partition during commit and then fail a job if a committer "went away" during that process. The existing MRv2 committer would return false here, as should the new EMR one. MRv1 and these S3A committers are resilient to failure here ... which is exactly what spark and Hadoop MR actually assume.

squito replied:
ok I took another look at this, and the key part that I was missing in the code I mentioned above is that it only matters when `customPartitionLocations` is defined.  I'd still like to understand that case better, but agree its not a show stopper.

squito commented on hadoop-cloud/src/hadoop-3/test/scala/org/apache/spark/internal/io/cloud/CommitterBindingSuite.scala:
```diff
@@ -0,0 +1,145 @@
+    }
+    val cause = ex.getCause
+    if (cause == null || !cause.isInstanceOf[IOException]
+      || !cause.getMessage.contains(PathOutputCommitProtocol.UNSUPPORTED)) {
```
nit: double indent continuation of the `if` condition

steveloughran replied:
added two spaces

squito commented:
style errors:

```
[error] /home/jenkins/workspace/SparkPullRequestBuilder@3/hadoop-cloud/src/hadoop-3/test/scala/org/apache/spark/internal/io/cloud/StubPathOutputCommitter.scala:82:35: No space before token =
[error] /home/jenkins/workspace/SparkPullRequestBuilder@3/hadoop-cloud/src/hadoop-3/test/scala/org/apache/spark/internal/io/cloud/StubPathOutputCommitter.scala:23:0: org.apache.hadoop.mapreduce. is in wrong order relative to org.apache.hadoop.mapreduce.lib.output..
[error] /home/jenkins/workspace/SparkPullRequestBuilder@3/hadoop-cloud/src/hadoop-3/test/scala/org/apache/spark/internal/io/cloud/CommitterBindingSuite.scala:28:0: org.apache.hadoop.mapreduce. is in wrong order relative to org.apache.hadoop.mapreduce.task.TaskAttemptContextImpl.
[error] /home/jenkins/workspace/SparkPullRequestBuilder@3/hadoop-cloud/src/hadoop-3/test/scala/org/apache/spark/internal/io/cloud/CommitterBindingSuite.scala:31:36: cloud should come before FileCommitProtocol.
```

you can run style checks locally with `dev/scalastyle`

steveloughran committed:
SPARK-23977: feedback + stylecheck

Removed the PathCommitterConstants entirely; those values used in tests were moved over there.

Reviewed the docs for correctness and
 * tweaked capitalisation/format in the new table
 * updated the reference list.

Change-Id: Ied893c3ac33afc8b1caf7e17273e6e50315f372a

SparkQA commented:
**[Test build #108519 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/108519/testReport)** for PR 24970 at commit [`b250612`](https://github.com/apache/spark/commit/b250612f27bb69f427079b1bead0b7d764d18415).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/13613/
Test PASSed.

SparkQA commented:
**[Test build #108519 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/108519/testReport)** for PR 24970 at commit [`b250612`](https://github.com/apache/spark/commit/b250612f27bb69f427079b1bead0b7d764d18415).
 * This patch **fails to generate documentation**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/108519/
Test FAILed.

vanzin commented:
```
[error] /home/jenkins/workspace/SparkPullRequestBuilder/hadoop-cloud/target/java/org/apache/spark/internal/io/cloud/BindingParquetOutputCommitter.java:4: error: reference not found
[error]  * output committer, and is intended to allow callers to use any {@link PathOutputCommitter},
[error]                                                                         ^
[error] /home/jenkins/workspace/SparkPullRequestBuilder/hadoop-cloud/target/java/org/apache/spark/internal/io/cloud/BindingParquetOutputCommitter.java:5: error: reference not found
[error]  * even if not a subclass of {@link ParquetOutputCommitter}.
[error]                                     ^
[info] Generating /home/jenkins/workspace/SparkPullRequestBuilder/target/javaunidoc/org/apache/spark/internal/io/cloud/CommitterBindingSuite.html...
[info] Generating /home/jenkins/workspace/SparkPullRequestBuilder/target/javaunidoc/org/apache/spark/internal/io/cloud/PathOutputCommitProtocol.html...
[error] /home/jenkins/workspace/SparkPullRequestBuilder/hadoop-cloud/target/java/org/apache/spark/internal/io/cloud/PathOutputCommitProtocol.java:22: error: reference not found
[error]  * @throws IOException when an unsupported dynamicPartitionOverwrite option is supplied.
[error]    ^
```

vanzin commented on core/src/main/scala/org/apache/spark/internal/io/HadoopMapReduceCommitProtocol.scala:
```diff
@@ -222,6 +222,7 @@ class HadoopMapReduceCommitProtocol(
 
   override def commitTask(taskContext: TaskAttemptContext): TaskCommitMessage = {
     val attemptId = taskContext.getTaskAttemptID
+    logDebug("Commit task")
```
This seems more noisy than useful. Either remove or at least add something to the message that lets you identify what's being committed.

But in general I'd refrain from adding this kind of logging at the task level.

steveloughran replied:
ok, now I'm confused. I'd put this in the subclass, and was told "pull it up". Which I did. Now I'm told "remove". Fine.  Done. But it would be good for both of us to have some consistency here: me for less wasted effort, you for less reviewing.

vanzin replied:
I've always complained about debug messages at task level. I'm pretty consistent about complaining about that because it makes debug logs less useful - it's ok with a few tasks but not with thousands.

steveloughran replied:
cut entirely

vanzin commented on docs/cloud-integration.md:
```diff
@@ -190,15 +212,50 @@ while they are still being written. Applications can write straight to the monit
+filesystem renames.
+
+Some object store connectors provide custom committers to commit tasks and
+jobs without using rename. In versions of Spark built with Hadoop-3.1 or later,
```
"Hadoop 3.1"

vanzin commented on docs/cloud-integration.md:
```diff
@@ -190,15 +212,50 @@ while they are still being written. Applications can write straight to the monit
+job commit are much faster, and task failures do not affect the result. 
+
+To switch to the S3A committers, use a version of Spark which includes the
+Hadoop-3.1+ binaries, and switch the committers through the following
```
"Hadoop 3.1+"

vanzin commented on docs/cloud-integration.md:
```diff
@@ -190,15 +212,50 @@ while they are still being written. Applications can write straight to the monit
+* [OpenStack Swift](https://hadoop.apache.org/docs/current/hadoop-openstack/index.html).
+* [Azure Blob Storage and Azure Datalake Gen 2](https://hadoop.apache.org/docs/current/hadoop-aws/tools/hadoop-aws/index.html).
+* [Azure Data Lake Gen 1](https://hadoop.apache.org/docs/current/hadoop-azure-datalake/index.html).
+* [Hadoop-AWS module (Hadoop 3-.x)](https://hadoop.apache.org/docs/current3/hadoop-aws/tools/hadoop-aws/index.html).
```
"3.x"

steveloughran replied:
done

vanzin commented on hadoop-cloud/src/hadoop-3/main/scala/org/apache/spark/internal/io/cloud/PathOutputCommitProtocol.scala:
```diff
@@ -0,0 +1,263 @@
+ * `org.apache.hadoop.mapreduce.lib.output.PathOutputCommitterFactory` factory
+ * API to create the committer.
+ *
+ * In `setupCommitter` the factory is identified and instantiated to;
```
"instantiated to"?

steveloughran replied:
removed the "to"

vanzin commented on hadoop-cloud/src/hadoop-3/main/scala/org/apache/spark/internal/io/cloud/PathOutputCommitProtocol.scala:
```diff
@@ -0,0 +1,263 @@
+   *         [[PathOutputCommitter]].
+   */
+  override protected def setupCommitter(context: TaskAttemptContext): PathOutputCommitter = {
+
```
nit: nix empty line

steveloughran replied:
done

vanzin commented on hadoop-cloud/src/hadoop-3/main/scala/org/apache/spark/internal/io/cloud/PathOutputCommitProtocol.scala:
```diff
@@ -0,0 +1,263 @@
+      taskContext: TaskAttemptContext,
+      absoluteDir: String,
+      ext: String): String = {
+
```
nit: nix empty line

steveloughran replied:
done

vanzin commented on hadoop-cloud/src/hadoop-3/main/scala/org/apache/spark/internal/io/cloud/PathOutputCommitProtocol.scala:
```diff
@@ -0,0 +1,263 @@
+  protected def buildFilename(
+      taskContext: TaskAttemptContext,
+      ext: String): String = {
+
```
nit: nix empty line

steveloughran replied:
done

vanzin commented on hadoop-cloud/src/hadoop-3/main/scala/org/apache/spark/internal/io/cloud/PathOutputCommitProtocol.scala:
```diff
@@ -0,0 +1,263 @@
+   * @param taskContext context
+   */
+  override def abortTask(taskContext: TaskAttemptContext): Unit = {
+    logDebug("Abort task")
```
See previous comments about unhelpful debug messages for task-level events.

(And you always have `logTrace`, too...)

steveloughran replied:
How about: logTrace with job/taskID? I know that spark JobIds are not as unique as MR ones, but it will give more context to the messages in a worker doing >1 job? Same for error messages generated when catching IOEs in job/task abort...task attempt IDs are potentially useful here

vanzin replied:
> logTrace with job/taskID

That sounds ok. Task ID is probably enough (if you mean `TaskContext.get.taskAttemptId`).

steveloughran replied:
I've cut all this completely. the s3a committers are fairly detailed if you ask them to log, and if you really need it, you can tell the AWS SDK to print the HTTP communications

vanzin commented on hadoop-cloud/src/hadoop-3/main/scala/org/apache/spark/internal/io/cloud/package.scala:
```diff
@@ -0,0 +1,98 @@
+   * Dataframe output to use the factory mechanism for writing data for
+   * all file formats.
+   */
+  val COMMITTER_BINDING_OPTIONS: Map[String, String] = Map(
```
This is not used anywhere and this package is internal... refer to previous discussion about this not being helpful in this form.

Basically, everything in this file is lost by the fact that this is ignored when generating API docs. So if there's useful info, it should be in the proper documentation.

steveloughran replied:
ok. There's enough in the docs, I'll cut it

vanzin commented on hadoop-cloud/src/hadoop-3/test/scala/org/apache/spark/internal/io/cloud/CommitterBindingSuite.scala:
```diff
@@ -0,0 +1,147 @@
+import org.apache.spark.internal.io.{cloud, FileCommitProtocol}
+
+/**
+ * Test committer binding logic.
```
pretty obvious from the class name.

steveloughran replied:
cut

vanzin commented on hadoop-cloud/src/hadoop-3/test/scala/org/apache/spark/internal/io/cloud/CommitterBindingSuite.scala:
```diff
@@ -0,0 +1,147 @@
+  }
+
+  /**
+   * Verify that the committer protocol can be serialized and that
```
Maybe have a better test name instead of a comment explaining the test?

steveloughran replied:
moved the text down; removed the comment.

vanzin commented on hadoop-cloud/src/hadoop-3/test/scala/org/apache/spark/internal/io/cloud/CommitterBindingSuite.scala:
```diff
@@ -0,0 +1,147 @@
+    assert("file:///tmp" === protocol.destination)
+  }
+
+  test("InstantiateNoDynamicPartitioning") {
```
We generally use proper sentences to describe tests.

steveloughran replied:
changed.

vanzin commented on hadoop-cloud/src/hadoop-3/test/scala/org/apache/spark/internal/io/cloud/StubPathOutputCommitter.scala:
```diff
@@ -0,0 +1,126 @@
+}
+
+/**
+ * Factory.
```
Pretty obvious from the class name.

steveloughran replied:
cut

vanzin commented on hadoop-cloud/src/hadoop-3/test/scala/org/apache/spark/internal/io/cloud/StubPathOutputCommitter.scala:
```diff
@@ -0,0 +1,126 @@
+    val key = OUTPUTCOMMITTER_FACTORY_SCHEME + "." + scheme
+    conf.set(key, classOf[StubPathOutputCommitterFactory].getName())
+  }
+
```
nit: too many empty lines

steveloughran replied:
fixed

vanzin reviewed (commented):
I think this just needs to be updated to address old comments that seem to have been forgotten.

steveloughran committed:
SPARK-23977: pull up as much as possible int HadoopMapReduceCommitProtocol

* log at trace there and include job/task ID
* fix up nits and typos
* Test suite moves to detail in test name

Change-Id: Ie7a931d52af2d0348aa02a2c66f8a5ae7b65e208
Testing: downstream tests against S3 Ireland + S3Guard; all seems good

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/13745/
Test PASSed.

SparkQA commented:
**[Test build #108659 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/108659/testReport)** for PR 24970 at commit [`58e1740`](https://github.com/apache/spark/commit/58e1740cc55e8d45da3d97939170d0e06b8d11e3).

SparkQA commented:
**[Test build #108659 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/108659/testReport)** for PR 24970 at commit [`58e1740`](https://github.com/apache/spark/commit/58e1740cc55e8d45da3d97939170d0e06b8d11e3).
 * This patch **fails to generate documentation**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/108659/
Test FAILed.

vanzin commented:
Still needs fixing the javadoc errors I pointed out previously...

steveloughran commented:
sorry, will do

steveloughran committed:
SPARK-23977

-remove traces
-fix scaladoc warnings
-docs 

Change-Id: If598e64ff803f14f122df813338c3814e4ddfb78

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/13955/
Test PASSed.

SparkQA commented:
**[Test build #108881 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/108881/testReport)** for PR 24970 at commit [`ee247d9`](https://github.com/apache/spark/commit/ee247d9c473f665791da1bbc84e1f61b35ff8dc1).

SparkQA commented:
**[Test build #108881 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/108881/testReport)** for PR 24970 at commit [`ee247d9`](https://github.com/apache/spark/commit/ee247d9c473f665791da1bbc84e1f61b35ff8dc1).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/108881/
Test PASSed.

squito commented on hadoop-cloud/src/hadoop-3/test/scala/org/apache/spark/internal/io/cloud/CommitterBindingSuite.scala:
```diff
@@ -0,0 +1,146 @@
+      out = new ObjectOutputStream(new FileOutputStream(serData))
+      out.writeObject(committer)
+      out.close
+      in = new ObjectInputStream(new FileInputStream(serData))
```
minor: this will be pretty small, right?  wouldn't it be easier to write to a `ByteArrayOutputStream` rather than messing with files?

```
val bytes = new ByteArrayOutputStream()
val out = new ObjectOutputStream(bytes)
...
val in = new ObjectInputStream(new ByteArrayInputSream(bytes.toByteArray()))
...
```

squito reviewed (approved):
I'd prefer the exception handling is changed to use supressed exceptions, rather than warn-and-discard, but not a big deal.

vanzin commented on core/src/main/scala/org/apache/spark/internal/io/HadoopMapReduceCommitProtocol.scala:
```diff
@@ -205,11 +206,28 @@ class HadoopMapReduceCommitProtocol(
+  /**
+   * Abort the job; log and ignore any IO exception thrown.
+   * This is invariably invoked in an exception handler; raising
+   * an exception here will lose the root cause of the failure.
```
>  isn't the right thing to do for those to catch the errors and use originalException.addSuppressed

That sounds better, but at the same time it's a change in all of the call sites, not here. We should have a bug open to do that (at which point we should remove these try..catch blocks).

vanzin commented on hadoop-cloud/src/hadoop-3/test/scala/org/apache/spark/internal/io/cloud/CommitterBindingSuite.scala:
```diff
@@ -0,0 +1,146 @@
+    job
+  }
+
+  test(" Verify that the committer protocol can be serialized and deserialized") {
```
Stray leading space

vanzin reviewed (commented):
I think this looks ok. I'll re-run the tests just in case and merge if they pass.

(I'll also rename the tests during merge since having them all say "Verify foo" is pretty redundant.)

vanzin commented:
retest this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/14201/
Test PASSed.

SparkQA commented:
**[Test build #109133 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/109133/testReport)** for PR 24970 at commit [`ee247d9`](https://github.com/apache/spark/commit/ee247d9c473f665791da1bbc84e1f61b35ff8dc1).

SparkQA commented:
**[Test build #109133 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/109133/testReport)** for PR 24970 at commit [`ee247d9`](https://github.com/apache/spark/commit/ee247d9c473f665791da1bbc84e1f61b35ff8dc1).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/109133/
Test PASSed.

vanzin commented:
Merging to master.

steveloughran commented:
thanks!

gaborgsomogyi commented:
This allows us to test streaming queries with S3 checkpoint location.
@steveloughran presume S3 "Zero Rename" committer is the thing which has to be set.

steveloughran commented:
yep, hadoop docs cover the setup; its a matter of configs. Use the directory one to start with, its the one we qualify
